File: Park/CodeTree/조건에 맞게 선택적으로 전진하는 DP/최대 점프 횟수.py
import sys

INT_MIN = -sys.maxsize

# 변수 선언 및 입력
n = int(input())
arr = list(map(int, input().split()))

# dp[i] : 위치 i에 도착할 때 가능한 최대 점프 횟수
dp = [0] * n

def initialize():
    # 초기값을 INT_MIN으로 설정
    for i in range(n):
        dp[i] = INT_MIN
    
    # 시작 위치의 초기조건 설정
    dp[0] = 0

# 초기조건 설정
initialize()

# 점화식을 통해 dp 값 채우기
for i in range(1, n):
    for j in range(i):
        if dp[j] == INT_MIN:
            continue
        if j + arr[j] >= i:
            dp[i] = max(dp[i], dp[j] + 1)

# 최댓값 계산
ans = max(dp)
print(ans)

File: jhyun/CodeTree/DP1/조건에 맞게 선택적으로 전진하는 DP/최대_점프_횟수.java
import java.util.*;
import java.io.*;

public class Main {
    static final int INT_MIN = Integer.MIN_VALUE;
    static final int MAX_N = 1000;

    static int n;
    static int[] dp = new int[MAX_N];
    static int[] arr = new int[MAX_N];

    static void init() {
        for (int i = 0; i < n; i++) {
            dp[i] = INT_MIN;
        }

        dp[0] = 0;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        init();

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j] == INT_MIN) {
                    continue;
                }

                if (j + arr[j] >= i) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
        }

        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, dp[i]);
        }
        System.out.println(ans);

    }
}